src/measurement_buffer.cpp

```diff
@@ -150,53 +129,46 @@ void MeasurementBuffer::BufferPCLCloud(const \
       return;
     }
 
-    point_cloud_ptr cld_global(new pcl::PointCloud<pcl::PointXYZ>);
-    geometry_msgs::TransformStamped tf_stamped = _buffer.lookupTransform(_global_frame, cloud.header.frame_id.c_str(), pcl_conversions::fromPCL(cloud.header).stamp);
-    tf::StampedTransform tf_transform;
-    tf::transformMsgToTF(tf_stamped.transform, tf_transform);
-    tf_transform.frame_id_ = _global_frame;
-    tf_transform.stamp_ = tf_stamped.header.stamp;
-    pcl_ros::transformPointCloud(cloud, *cld_global, tf_transform);
-    cld_global->header.stamp = cloud.header.stamp;
+    // transform the cloud in the global frame
+    point_cloud_ptr cld_global(new sensor_msgs::PointCloud2());
+    geometry_msgs::TransformStamped tf_stamped = _buffer.lookupTransform(_global_frame, cloud.header.frame_id, cloud.header.stamp);
+    tf2::doTransform (cloud, *cld_global, tf_stamped);
+
+    pcl::PCLPointCloud2::Ptr cloud_pcl (new pcl::PCLPointCloud2 ());
+    pcl::PCLPointCloud2::Ptr cloud_filtered (new pcl::PCLPointCloud2 ());
 
-    // if user wants to use a voxel filter
+    pcl_conversions::toPCL(*cld_global, *cloud_pcl);
+
+    // remove points that are below or above our height restrictions, and in the same time, remove NaNs
+    // and if user wants to use it, combine with a voxel filter
     if ( _voxel_filter )
     {
-      // remove nans because they show things down
-      point_cloud_ptr cld_no_nan(new pcl::PointCloud<pcl::PointXYZ>);
-      std::vector<int> indices;
-      pcl::removeNaNFromPointCloud(*cld_global, *cld_no_nan, indices);
-
-      // minimize information needed to process
-      cld_global->clear();
-      pcl::ApproximateVoxelGrid<pcl::PointXYZ> sor1;
-      sor1.setInputCloud (cld_no_nan);
-      sor1.setLeafSize ((float)_voxel_size,
-                        (float)_voxel_size,
-                        (float)_voxel_size);
-      sor1.filter (*cld_global);
+        pcl::VoxelGrid<pcl::PCLPointCloud2> sor;
+        sor.setInputCloud (cloud_pcl);
+        sor.setFilterFieldName("z");
```

Comment:
This height filter is probably the reason you get a performance increase, since this is a relatively heavy operation.

Reply:
+ having to deal with simpler conversions